File: App/app/src/main/java/com/cinema/ticket_booking/ui/booking/SelectShowtimeViewModel.java
package com.cinema.ticket_booking.ui.booking;

import androidx.lifecycle.*;
import com.cinema.ticket_booking.data.model.response.*;
import com.cinema.ticket_booking.data.repository.*;
import com.cinema.ticket_booking.util.Resource;
import java.text.SimpleDateFormat;
import java.util.*;
import dagger.hilt.android.lifecycle.HiltViewModel;
import javax.inject.Inject;

@HiltViewModel
public class SelectShowtimeViewModel extends ViewModel {
    private final ShowtimeRepository showtimeRepo;
    private final CinemaRepository cinemaRepo;

    private final MutableLiveData<Resource<List<ShowtimeResponse>>> showtimes = new MutableLiveData<>();
    private final MutableLiveData<Resource<List<CinemaResponse>>> cinemas = new MutableLiveData<>();
    private final MutableLiveData<String> selectedDate = new MutableLiveData<>();
    private final MutableLiveData<String> selectedCinemaId = new MutableLiveData<>();
    private String movieId;

    // Booking state shared sang SelectSeat
    public static String pendingShowtimeId;
    public static String pendingMovieTitle;
    public static String pendingShowtimeTime;
    public static String pendingCinemaName;
    public static String pendingMoviePoster;
    public static String pendingShowDate;

    @Inject
    public SelectShowtimeViewModel(ShowtimeRepository showtimeRepo, CinemaRepository cinemaRepo) {
        this.showtimeRepo = showtimeRepo;
        this.cinemaRepo = cinemaRepo;
        selectedDate.setValue(todayDate());
    }

    public LiveData<Resource<List<ShowtimeResponse>>> getShowtimes() {
        return showtimes;
    }

    public LiveData<Resource<List<CinemaResponse>>> getCinemas() {
        return cinemas;
    }

    public LiveData<String> getSelectedDate() {
        return selectedDate;
    }

    public String getSelectedCinemaId() {
        return selectedCinemaId.getValue();
    }

    public void loadCinemas() {
        cinemaRepo.getCinemas(null).observeForever(cinemas::setValue);
    }

    public void selectDate(String date) {
        selectedDate.setValue(date);
        loadShowtimes(movieId);
    }

    public void selectCinema(String cinemaId) {
        selectedCinemaId.setValue(cinemaId);
        loadShowtimes(movieId);
    }

    public void loadShowtimes(String movieId) {
        if (movieId != null) {
            this.movieId = movieId;
        }
        String date = selectedDate.getValue();
        String cinema = selectedCinemaId.getValue();
        // Cần ít nhất movieId hoặc cinemaId để gọi API
        if (this.movieId == null && cinema == null) return;
        showtimeRepo.getShowtimes(this.movieId, cinema, date).observeForever(showtimes::setValue);
    }

    public List<String> getNext7Days() {
        List<String> dates = new ArrayList<>();
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd", Locale.getDefault());
        Calendar cal = Calendar.getInstance();
        for (int i = 0; i < 7; i++) {
            dates.add(sdf.format(cal.getTime()));
            cal.add(Calendar.DAY_OF_MONTH, 1);
        }
        return dates;
    }

    private String todayDate() {
        return new SimpleDateFormat("yyyy-MM-dd", Locale.getDefault()).format(new Date());
    }
}

File: App/app/src/main/java/com/cinema/ticket_booking/ui/booking/SelectShowtimeViewModel.kt
package com.cinema.ticket_booking.ui.booking

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.cinema.ticket_booking.data.model.response.CinemaResponse
import com.cinema.ticket_booking.data.model.response.ShowtimeResponse
import com.cinema.ticket_booking.data.repository.CinemaRepository
import com.cinema.ticket_booking.data.repository.ShowtimeRepository
import com.cinema.ticket_booking.util.Resource
import dagger.hilt.android.lifecycle.HiltViewModel
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale
import javax.inject.Inject

@HiltViewModel
class SelectShowtimeViewModel @Inject constructor(
    private val showtimeRepo: ShowtimeRepository,
    private val cinemaRepo: CinemaRepository
) : ViewModel() {

    private val _showtimes = MutableLiveData<Resource<List<ShowtimeResponse>>>()
    val showtimes: LiveData<Resource<List<ShowtimeResponse>>> = _showtimes

    private val _cinemas = MutableLiveData<Resource<List<CinemaResponse>>>()
    val cinemas: LiveData<Resource<List<CinemaResponse>>> = _cinemas

    private val _selectedDate = MutableLiveData<String>()
    val selectedDate: LiveData<String> = _selectedDate

    private val _selectedCinemaId = MutableLiveData<String?>()
    val selectedCinemaId: LiveData<String?> = _selectedCinemaId

    private var movieId: String? = null

    companion object {
        @JvmField
        var pendingShowtimeId: String? = null
        @JvmField
        var pendingMovieTitle: String? = null
        @JvmField
        var pendingShowtimeTime: String? = null
        @JvmField
        var pendingCinemaName: String? = null
        @JvmField
        var pendingMoviePoster: String? = null
        @JvmField
        var pendingShowDate: String? = null
    }

    init {
        _selectedDate.value = todayDate()
    }

    fun getSelectedCinemaId(): String? {
        return _selectedCinemaId.value
    }

    fun loadCinemas() {
        cinemaRepo.getCinemas(null).observeForever { _cinemas.value = it }
    }

    fun selectDate(date: String) {
        _selectedDate.value = date
        loadShowtimes(movieId)
    }

    fun selectCinema(cinemaId: String?) {
        _selectedCinemaId.value = cinemaId
        loadShowtimes(movieId)
    }

    fun loadShowtimes(movieId: String?) {
        if (movieId != null) {
            this.movieId = movieId
        }
        val date = _selectedDate.value
        val cinema = _selectedCinemaId.value
        // Cần ít nhất movieId hoặc cinemaId để gọi API
        if (this.movieId == null && cinema == null) return
        showtimeRepo.getShowtimes(this.movieId, cinema, date).observeForever { _showtimes.value = it }
    }

    val next7Days: List<String>
        get() {
            val dates = ArrayList<String>()
            val sdf = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault())
            val cal = Calendar.getInstance()
            for (i in 0..6) {
                dates.add(sdf.format(cal.time))
                cal.add(Calendar.DAY_OF_MONTH, 1)
            }
            return dates
        }

    private fun todayDate(): String {
        return SimpleDateFormat("yyyy-MM-dd", Locale.getDefault()).format(Date())
    }
}
